pilot/pkg/xds/ads.go

```diff
@@ -173,7 +173,7 @@ func (s *DiscoveryServer) processRequest(req *discovery.DiscoveryRequest, con *C
 	// It can happen when `processRequest` comes after push context has been updated(s.initPushContext),
 	// but proxy's SidecarScope has been updated(s.computeProxyState -> SetSidecarScope) due to optimizations that skip sidecar scope
 	// computation.
-	if con.proxy.SidecarScope != nil && con.proxy.SidecarScope.Version != request.Push.PushVersion {
+	if len(con.proxy.ServiceTargets) == 0 || (con.proxy.SidecarScope != nil && con.proxy.SidecarScope.Version != request.Push.PushVersion) {
```

Comment:
Hmm, it seems like if we are getting the _wrong pod_ does that mean we could get some services and just the wrong one? Or will it always be empty?

One concern here is its valid to have 0 services and now we need to recompute every time which could be pretty expensive?

Reply:
> I'm not really sure that this fix is correct, ti catches the case where no service have matched this proxy, but other cases could appear.

100% agree. I was also not comfortable with this fix. But just wanted to get others thoughts as well on this. I do not think there is easy way to identify when to recompute service targets - One way is always recompute but that is expensive to do

> what I'm not certain is if this happened even with the fix in https://github.com/istio/istio/pull/56502, what I mean is, we've stopped experiencing this issues with https://github.com/istio/istio/pull/56502, if that's enough, then why make these changes?

Good to know that fix helped you. I will remove these changes and leave the metrics part in this PR